contracts/Removal.sol

```diff
@@ -80,57 +110,137 @@ contract Removal is
       data,
       (BatchMintRemovalsData)
     );
-    for (uint256 i = 0; i < ids.length; i++) {
-      if (_tokenIdExists[ids[i]]) {
-        revert TokenIdExists(); // todo can the duplicate token id be reported here?
+    uint256 projectId = decodedData.projectId;
+    uint256 scheduleStartTime = decodedData.scheduleStartTime;
+    bool list = decodedData.list;
+    address marketAddress = decodedData.marketAddress;
+    if (projectId == 0) {
+      revert InvalidProjectId({projectId: projectId});
+    }
+    if (
+      scheduleStartTime < UNIX_EPOCH_2010 || scheduleStartTime > UNIX_EPOCH_2050
+    ) {
+      revert InvalidScheduleStartTime({startTime: scheduleStartTime});
+    }
+    uint256 newTokenIdCounter = tokenIdCounter;
+    for (uint256 i = 0; i < idsLength; i++) {
+      uint256 id = ids[i];
+      if (_tokenIdExists[id]) {
+        revert TokenIdExists({tokenId: id});
       }
-      _tokenIdExists[ids[i]] = true;
-      indexToTokenId[_tokenIdCounter] = ids[i];
-      _tokenIdCounter += 1;
+      _removalIdToProjectId[id] = projectId;
+      _tokenIdExists[id] = true;
+      indexToTokenId[newTokenIdCounter] = id;
+      newTokenIdCounter += 1;
     }
+    uint256 firstId = ids[0];
+    tokenIdCounter = newTokenIdCounter;
+    _projectIdToScheduleData[projectId] = ScheduleData(
+      scheduleStartTime,
+      firstId.supplierAddress(), // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
+      firstId.methodology(), // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
+      firstId.methodologyVersion() // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
+    );
     super.mintBatch(to, ids, amounts, data);
-
     setApprovalForAll(to, _msgSender(), true); // todo look at vesting contract for potentially better approach
-    if (decodedData.list) {
-      safeBatchTransferFrom(to, decodedData.marketAddress, ids, amounts, data);
+    if (list) {
+      safeBatchTransferFrom(to, marketAddress, ids, amounts, data);
     }
   }
 
+  function mint(
+    address,
+    uint256,
+    uint256,
+    bytes calldata
+  ) public pure override {
+    revert("Removal: ERC 1155 mint disabled");
+  }
+
   /**
-   * @dev used to initiate a sale of removals by transferring the removals to the market contract
+   * @dev used to list removals for sale by transferring the removals to the market contract
+   *
+   * ### Requirements:
+   *  - all removals being listed must belong to the same project id.
    */
   function safeBatchTransferFrom(
-    address _from,
-    address _to,
-    uint256[] memory _ids,
-    uint256[] memory _amounts,
-    bytes memory _data
+    address from,
+    address to,
+    uint256[] memory ids,
+    uint256[] memory amounts,
+    bytes memory data
   ) public override {
+    // todo do we add any validation to enforce that all removals in batch belong to the same project id?
```

Comment:
this would involve decoding the project id and making a comparison for each removal in the batch in a function that is already very costly... how do you feel about leaving this todo here and not including this work as part of this PR, at least?